CIP-0020: Canonical CBOR serialisation (#2)
vacuumlabs/CIPs: CIP-0020/CIP-0020.md

---
CIP: 0016
Title: Canonical CBOR serialisation
Authors: Gabriel Kerekes <gabriel.kerekes@vacuumlabs.com>, Rafael Korbas <rafael.korbas@vacuumlabs.com>
Comments-URI: https://forum.cardano.org/t/stake-pool-cold-keys-hd-derivation/43360
Status: Draft
Type: Standards
Created: 2021-06-14
License: CC-BY-4.0
---

## Abstract

This CIP defines a canonical format for CBOR serialisation in Cardano by extending the suggestions for canonical CBOR from the [CBOR specification RFC](https://datatracker.ietf.org/doc/html/rfc7049#section-3.9).

## Motivation

> It’s desirable to have interoperability between different transaction signing tools (e.g. cardano-cli, HW wallets, cardano-hw-cli), some of which may need to serialise the transaction by itselves (especially HW wallets, due to security/performance considerations). By interoperability we mean that given the same inputs (transaction parameters) the tools would produce the same outputs (e.g. transaction hash).

Reviewer: ```suggestion
While Cardano nodes are quite lenient in accepting minor variations in transaction serialization format allowed by the CDDL specification, certain tools might require additional restrictions ensuring interoperability. This is especially important for tools working with hardware wallets which need to reserialize transactions (due to their memory limitations and certain security considerations). In order to ensure consistency between the transaction body and all the witness signatures, we need to specify a canonical format for CBOR serialization.
```

> This is particularly relevant with the introduction of multisig. If the party initially creating the transaction uses e.g. cardano-cli to create the transaction, then other involved multisig parties might not be able to sign the transaction using HW wallets, because HW wallets might order some elements in the transaction differently thus they would come up with a different transaction hash. The same applies when the transaction is initially created by a HW wallet and requires signatures from parties using cardano-cli.

Reviewer: ```suggestion
This is particularly relevant with the introduction of multisig. If the party initially creating the transaction uses e.g. cardano-cli to create the transaction, then other involved multisig parties might not be able to sign the transaction using HW wallets, because HW wallets might order some elements in the transaction differently, ending up with a different transaction hash. The same applies when the transaction is initially created by a HW wallet and requires signatures from parties using cardano-cli or even some custom third-party tool.
```

> For example, in order to sign a transaction with a HW wallet, cardano-hw-cli takes the serialised transaction body from cardano-cli, parses it and transforms it into input parameters for the HW wallet. Without canonical CBOR serialisation it might happen that the HW wallet will calculate a different transaction hash than cardano-cli and the signature returned would thus be invalid for the transaction created by cardano-cli.

Reviewer: ```suggestion
For example, in order to sign a transaction with a HW wallet, cardano-hw-cli takes the serialised transaction body from cardano-cli, parses it and transforms it into input parameters for the HW wallet. Without canonical CBOR serialisation it might happen that the HW wallet will calculate a different transaction hash than cardano-cli and the returned witness signature would thus be invalid for the transaction created by cardano-cli.
```

Another motivation for this CIP is that without canonical CBOR serialisation, specifically canonical map key order, HW wallets would not be able to verify that multi-asset policies and asset_names are included only once in a given output or policy respectively, as they don't have the memory to store all previously passed tokens for the sake of deduplication. This might then cause the user to sign a transaction with the same token included twice, although at the protocol level, only one use of the token would be accounted for.

## Specification

This specification is a modification of the canonical CBOR suggestions taken from the [CBOR specification RFC](https://datatracker.ietf.org/doc/html/rfc7049#section-3.9). Parts not mentioned in this CIP should follow the canonical CBOR suggestions from the CBOR RFC.

### Map key ordering

> Keys in a map should be ordered lexicographically based on their real, logical value, not their CBOR representation. If the keys are of different types then the order described below should be maintained.

Reviewer: This is too imprecise. For instance, what does "lexicographical" mean for tx body keys? We use numerical value, not lexicographical ordering of their string representations.

Author: Good point. I've split it into number types and strings.

**Ordering of keys with different types**

The order of types is based on the CBOR major type’s 3 bit header identificator:

1. unsigned integer - major type 0
2. negative integer - major type 1
3. byte string - major type 2
4. text string - major type 3
5. floating-point numbers - major type 7

> I.e. first come the uints, then the ints, byte strings, text strings and floating-point numbers.

Reviewer: I am afraid this would clash with what cardano-cli authors consider "logical". I.e. if we have only integers as keys (including negative), why include the positive ones first?...

Author: Hm. That makes sense. I just don't know what the right approach is then. What if floating-point numbers came into the mix? Perhaps numbers should come first and be ordered by size no matter the type e.g.:

```
-123, -20.7, 30, 30.5
```

Would that make sense? And after that would come bytes and then strings. I.e.:

```
-123, -20.7, 30, 30.5, h'1010', h'20', 'aa', 'b'
```

@refi93 what do you think?

Reviewer: I think this decision should be left to IOHK. I don't think floating point numbers will ever be used (notice how pool margin is dealt with).

But whatever is written here, it must correspond to tx body map key ordering.

Arrays, maps and tagged items are left out of this specification as including them might unnecessarily complicate the specification or its implementation. Also, these elements aren’t currently used as keys in maps and are unlikely to be used in the future.

**Examples of correctly ordered keys**

- `"aa", "b", "c", "cccc", "d"`
- `h'0000', h'01', h'02020202', h'03'`
- `1, 2, 3, 4, 100, 1000`
- `1, 100, -100, -200, h'01', h'0202020202', "aa", "b", 0.7, 2.4`

## Rationale

### Why not follow the canonical CBOR spec completely?

The canonical CBOR specification orders the keys based on their CBOR representations with the rather weird caveat that shorter keys come first regardless of their lexicographic ordering. Lexicographic ordering is more natural and it might be simpler for many applications to sort based on the real, logical value of the key rather than the serialised representation.

### Keys with mixed types

We mustn’t forget that the CBOR map keys can be of different types. The type ordering can be chosen arbitrarily but it seems most natural to choose the ordering based on the CBOR types’ 3 bit header identificator.

## Backwards compatibility

> Canonical CBOR sation defined in this CIP is not compatible with canonical CBOR serialisation defined by the CBOR RFC. We recommend that all tools migrate to the canonical CBOR serialisation defined by this CIP in order to ensure interoperability across the ecosystem.

Reviewer: What's "sation"?

Author: Serialisation. Must have happened when copying from the doc. Thanks.

---
CIP: 0016
Title: Canonical CBOR serialisation
Authors: Gabriel Kerekes <gabriel.kerekes@vacuumlabs.com>, Rafael Korbas <rafael.korbas@vacuumlabs.com>, Jan Mazak <jan.mazak@vacuumlabs.com>

> Comments-URI: https://forum.cardano.org/t/stake-pool-cold-keys-hd-derivation/43360

Reviewer: This seems wrong.

Author: Hm, I thought I've removed it 🤦

This is particularly relevant with the introduction of multisig. If the party initially creating the transaction uses e.g. cardano-cli to create the transaction, then other involved multisig parties might not be able to sign the transaction using HW wallets, because HW wallets might order some elements in the transaction differently, ending up with a different transaction hash. The same applies when the transaction is initially created by a HW wallet and requires signatures from parties using cardano-cli or even some custom third-party tool.

For example, in order to sign a transaction with a HW wallet, cardano-hw-cli takes the serialised transaction body from cardano-cli, parses it and transforms it into input parameters for the HW wallet. Without canonical CBOR serialisation it might happen that the HW wallet will calculate a different transaction hash than cardano-cli and the returned witness signature would thus be invalid for the transaction created by cardano-cli.

> Another motivation for this CIP is that without canonical CBOR serialisation, specifically canonical map key order, HW wallets would not be able to verify that multi-asset policies and asset_names are included only once in a given output or policy respectively, as they don't have the memory to store all previously passed tokens for the sake of deduplication. This might then cause the user to sign a transaction with the same token included twice, although at the protocol level, only one use of the token would be accounted for.

Reviewer: ```suggestion
Another motivation for this CIP is that without canonical CBOR serialisation, specifically canonical map key order, HW wallets would not be able to verify that multi-asset policies and asset_names are included only once in a given output or policy respectively, as they don't have enough memory to store all the previously passed tokens for the sake of deduplication. This might then cause the user to sign a transaction with the same token included twice, although at the protocol level, only one use of the token would be accounted for.
```
